File: Daily/564. Find the Closest Palindrome.js
/**
 * @param {string} n
 * @return {string}
 */
var nearestPalindromic = function (n) {
    const len = n.length;
    const i = len % 2 === 0 ? len / 2 - 1 : Math.floor(len / 2);
    const firstHalf = BigInt(n.substring(0, i + 1));

    const possibilities = [];
    possibilities.push(halfToPalindrome(firstHalf, len % 2 === 0));
    possibilities.push(halfToPalindrome(firstHalf + 1n, len % 2 === 0));
    possibilities.push(halfToPalindrome(firstHalf - 1n, len % 2 === 0));
    possibilities.push(BigInt(10) ** BigInt(len - 1) - 1n);
    possibilities.push(BigInt(10) ** BigInt(len) + 1n);

    let diff = BigInt(Number.MAX_VALUE),
        res = 0n;
    const nl = BigInt(n);
    for (const cand of possibilities) {
        if (cand === nl) continue;
        const currentDiff = cand > nl ? cand - nl : nl - cand;
        if (currentDiff < diff) {
            diff = currentDiff;
            res = cand;
        } else if (currentDiff === diff) {
            res = cand < res ? cand : res;
        }
    }

    return res.toString();
};

var halfToPalindrome = function (left, even) {
    let res = left;
    if (!even) left = left / 10n;
    while (left > 0n) {
        res = res * 10n + (left % 10n);
        left = left / 10n;
    }
    return res;
};

console.log(nearestPalindromic("807045053224792883"));

File: Daily/564. Find the Closest Palindrome.java
class Solution {

    public String nearestPalindromic(String n) {
        int len = n.length();
        int i = len % 2 == 0 ? len / 2 - 1 : len / 2;
        long firstHalf = Long.parseLong(n.substring(0, i + 1));

        List<Long> possibilities = new ArrayList<>();
        possibilities.add(halfToPalindrome(firstHalf, len % 2 == 0));
        possibilities.add(halfToPalindrome(firstHalf + 1, len % 2 == 0));
        possibilities.add(halfToPalindrome(firstHalf - 1, len % 2 == 0));
        possibilities.add((long) Math.pow(10, len - 1) - 1);
        possibilities.add((long) Math.pow(10, len) + 1);

        long diff = Long.MAX_VALUE, res = 0, nl = Long.parseLong(n);
        for (long cand : possibilities) {
            if (cand == nl)
                continue;
            if (Math.abs(cand - nl) < diff) {
                diff = Math.abs(cand - nl);
                res = cand;
            } else if (Math.abs(cand - nl) == diff) {
                res = Math.min(res, cand);
            }
        }

        return String.valueOf(res);
    }

    private long halfToPalindrome(long left, boolean even) {
        long res = left;
        if (!even)
            left = left / 10;
        while (left > 0) {
            res = res * 10 + (left % 10);
            left /= 10;
        }
        return res;
    }
}